sdk/identity/azure-identity/src/default_azure_credential.cpp

```diff
@@ -38,17 +43,46 @@ DefaultAzureCredential::DefaultAzureCredential(
             "is the better fit for the application.");
 
   // Creating credentials in order to ensure the order of log messages.
-  auto const envCred = std::make_shared<EnvironmentCredential>(options);
-  auto const wiCred = std::make_shared<WorkloadIdentityCredential>(options);
-  auto const azCliCred = std::make_shared<AzureCliCredential>(options);
-  auto const managedIdentityCred = std::make_shared<ManagedIdentityCredential>(options);
+  ChainedTokenCredential::Sources miSources;
+  {
+    miSources.emplace_back(std::make_shared<EnvironmentCredential>(options));
+    miSources.emplace_back(std::make_shared<WorkloadIdentityCredential>(options));
+
+    constexpr auto envVarName = "AZURE_TOKEN_CREDENTIALS";
+    const auto envVarValue = Environment::GetVariable(envVarName);
+
+    const auto isProd = StringExtensions::LocaleInvariantCaseInsensitiveEqual(envVarValue, "prod");
+    const auto logMsg = GetCredentialName() + ": '" + envVarName + "' environment variable is "
+        + (envVarValue.empty() ? "not set" : ("set to '" + envVarValue + "'"))
+        + ", therefore AzureCliCredential will " + (isProd ? "NOT " : "")
+        + "be included in the credential chain.";
+
+    if (isProd)
+    {
+      IdentityLog::Write(IdentityLog::Level::Verbose, logMsg);
+    }
+    else if (
+        envVarValue.empty()
+        || StringExtensions::LocaleInvariantCaseInsensitiveEqual(envVarValue, "dev"))
+    {
+      IdentityLog::Write(IdentityLog::Level::Verbose, logMsg);
+      miSources.emplace_back(std::make_shared<AzureCliCredential>(options));
+    }
+    else
+    {
+      throw AuthenticationException(
+          GetCredentialName() + ": Invalid value '" + envVarValue + "' for the '" + envVarName
+          + "' environment variable. Allowed values are 'dev' and 'prod' (case insensitive). "
+            "It is also allowed to not have the environment variable defined.");
+    }
+
+    miSources.emplace_back(std::make_shared<ManagedIdentityCredential>(options));
```

Comment:
Am I misreading this or is the `dev` chain environment -> workload identity -> Azure CLI -> managed identity?

Reply:
Yes, and you are right, that has been a long problem which will be fixed soon in #6573.